source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp

```diff
@@ -0,0 +1,340 @@
+{
+    return std::wstring();
+}
+
+std::wstring DbgHelpStackWalkCallbackPrivate::GetParameterName(DWORD num)
+{
+    return std::wstring();
+}
+
+std::shared_ptr<void> DbgHelpStackWalkCallbackPrivate::GetParameterValue(DWORD num)
+{
+    return std::shared_ptr<void>();
+}
+
+CComPtr<IDiaSymbol> DbgHelpStackWalkCallbackPrivate::GetFunctionSymbol()
+{
+    CComPtr<IDiaSymbol> symbol;
+    HRESULT rv = m_mod_info->m_module_symbol_session->findSymbolByVA(GetProgramCounter(), SymTagFunction, &symbol);
+    return symbol;
+}
+
+ULONGLONG DbgHelpStackWalkCallbackPrivate::GetProgramCounter()
+{
+    ULONGLONG pc;
+    m_frame->get_registerValue(CV_REG_EIP, &pc);
+    return pc;
+}
+
+// TODO: RFC
+void DbgHelpStackWalkCallbackPrivate::EnumerateParameters()
+{
+    if (m_mod_info->m_module_symbol_session == nullptr)
+    {
+        /*
+         * TODO: Look at callee asm to guess number of parameters. The get_lengthParams method 
+         *       always returns 0 when no debug symbols are known...
+         * Temp workaround: Assume 4 parameters (this is what the old implementation did).
+         * -- Warepire
+         */
+        m_unsure++;
+        for (DWORD i = 0; i < 4; i++)
+        {
+            m_param_info.emplace_back(ParamInfo{ DbgHelpArgType::Int, L"int", { static_cast<WCHAR>(L'a' + i) }, nullptr, nullptr, 4, i * 4 });
+        }
+    }
+    else
+    {
+        auto symbol = GetFunctionSymbol();
+        CComPtr<IDiaEnumSymbols> enum_symbols;
+        HRESULT rv = symbol->findChildren(SymTagNull, nullptr, nsNone, &enum_symbols);
+        if (enum_symbols == nullptr)
+        {
+            return;
+        }
+        IDiaSymbol* sym_info;
+        ULONG num_fetched;
+
+        /*
+         * This will be incremented for each found symbol, with it's size.
+         */
+        DWORD offset = 0;
+        for (DWORD count = 0; (enum_symbols->Next(1, &sym_info, &num_fetched) == S_OK); count++)
+        {
+            if (num_fetched != 1)
+            {
+                sym_info->Release();
+                continue;
+            }
+            CComPtr<IDiaSymbol> type_info;
+            if (sym_info->get_type(&type_info) != S_OK)
+            {
+                /*
+                 * This will filter out most non-argument children in the function
+                 */
+                sym_info->Release();
+                continue;
+            }
+            DWORD symtag;
+            if (sym_info->get_symTag(&symtag) != S_OK || symtag == SymTagCallSite)
+            {
+                sym_info->Release();
+                continue;
+            }
+            DWORD typetag;
+            if (type_info->get_symTag(&typetag) != S_OK)
+            {
+                sym_info->Release();
+                continue;
+            }
+            ULONGLONG length = 0;
+            if (type_info->get_length(&length) != S_OK)
+            {
+                sym_info->Release();

```

Comment:
Sadly, these throws exceptions when you stuff them in `CComPtr` classes. Because when you reach the end of a DIA enum `sym_info` contains invalid data.